crates/wash-lib/src/run/mod.rs

```diff
@@ -0,0 +1,95 @@
+use crate::registry::get_oci_artifact;
+use crate::run::ctx::Ctx;
+use anyhow::{Context, Result};
+use wasmtime::component::Component;
+use wasmtime::{Config, Engine};
+use wasmtime_wasi::{WasiCtx, WasiView};
+
+pub use crate::run::ctx::CtxBuilder;
+
+pub use wasmtime_wasi::{DirPerms, FilePerms};
+
+pub mod ctx;
+
+/// Runs workloads capable of leveraging the local host it runs on.
+///
+/// Access to the host's resources can be strictly controlled.
+/// Follows the principle of least privilege, which means all is denied by
+/// default.
+pub struct LocalRuntime {
+    engine: Engine,
+}
+
+impl LocalRuntime {
+    pub fn new() -> Result<Self> {
+        let mut config = Config::new();
+
+        // Try to read settings from default location
+        if let Err(e) = config.cache_config_load_default() {
+            tracing::warn!(err = ?e, "Failed to load wasm cache");
+        }
+
+        config.wasm_component_model(true);
+        config.async_support(true);
+
+        let engine = Engine::new(&config)?;
+        anyhow::Ok(Self { engine })
+    }
+
+    pub async fn run(&self, context: Ctx) -> Result<()> {
+        let artifact = get_oci_artifact(
+            context.reference.clone(),
+            context.oci_cache_file.clone(),
+            context.oci_pull_options.clone(),
+        )
+        .await
+        .context("failed to pull the component")?;
+
+        let component = Component::new(&self.engine, &artifact)
+            .context("failed to build the pulled component")?;
+
+        //NB(raskyld): Not sure if the `data` stored needs to be owned by the Store itself.
+        let mut store = wasmtime::Store::new(&self.engine, State::new(context));
+        let mut linker = wasmtime::component::Linker::new(&self.engine);
+
+        wasmtime_wasi::bindings::Command::add_to_linker(&mut linker, |state: &mut State| state)?;
+
+        //TODO(raskyld): add http host funcs
+
+        let (instance, _) =
+            wasmtime_wasi::bindings::Command::instantiate_async(&mut store, &component, &linker)
+                .await
+                .context("failed to instantiate the compiled component")?;
+
+        instance
+            .wasi_cli_run()
+            .call_run(&mut store)
+            .await
+            .context("failed to run the component")?
+            .map_err(|_| anyhow::anyhow!("the component has been run but returned an error"))
```

Comment:
Could we reuse any functionality from `wash-lib/src/plugin/subcommand.rs` when it comes to wasmtime Ctx / executing the run function? @thomastaylor312 might have some thoughts. I know linking is different but it might help here since in both cases we're calling `wasi:cli/run`

Reply:
There may be overlaps indeed, especially since a `wasi:cli/run` runtime could be used to run `subcommand` (as long as it gives control over the linking step).

When I looked at the plugin system, the main problem that motivated me to re-develop the abstractions is that I needed my API to give control over host permissions and ATM, the permissions are specifically made for the case of running sub commands.

Maybe we should write a `cli` + `http` runner whose API let the caller customise the permissions and extend the linking process and then use it in `Subcommand`? 🤔

Reply:
That makes sense. I know we wanted plugins to possibly have HTTP access, and eventually have NATS access in order to do things like query the state of the lattice and/or manipulate lattice things like starting components. I'll see if Taylor has thoughts here since the plugins were meant to coursely allow permissions.

I don't want to make the code more complicated re: re-use, but I want to make sure we aren't maintaining two separate wasi:cli/run implementations 😄

Reply:
Yeah I don't really know if there would be a lot we could reuse. Generally you can reuse the `Engine` but since this is a cli tool, you can't simultaneously be running plugin commands and the run command, so it doesn't really gain us anything. I'll take a look in just a second though